File: app/src/main/java/com/example/nasmovie/ui/FavoritesFragment.java
package com.example.nasmovie.ui;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.nasmovie.R;
import com.example.nasmovie.data.model.Movie;
import com.example.nasmovie.data.repository.MovieRepository;
import com.example.nasmovie.ui.adapter.FavoriteAdapter;
import com.google.android.material.button.MaterialButton;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Set;

/**
 * 收藏 Fragment
 */
public class FavoritesFragment extends Fragment implements
        FavoriteAdapter.OnMovieClickListener,
        FavoriteAdapter.OnMovieLongClickListener {

    private com.example.nasmovie.view.NasToolbar toolbar;
    private RecyclerView recyclerView;
    private LinearLayout emptyView;
    private View bottomActions;
    private MaterialButton btnDelete;
    private EditText etSearch;

    private FavoriteAdapter adapter;
    private MovieRepository repository;
    private List<Movie> allFavorites = new ArrayList<>();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_favorites, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initViews(view);
        initData();
    }

    private void initViews(View view) {
        toolbar = view.findViewById(R.id.toolbar);
        recyclerView = view.findViewById(R.id.recycler_view_favorites);
        emptyView = view.findViewById(R.id.empty_view);
        bottomActions = view.findViewById(R.id.bottom_actions);
        btnDelete = view.findViewById(R.id.btn_delete);
        etSearch = view.findViewById(R.id.et_search);

        // 设置 Toolbar
        if (toolbar != null) {
            toolbar.setTitle("我的收藏");
            toolbar.setShowBack(false);
            AppCompatActivity activity = (AppCompatActivity) getActivity();
            if (activity != null) {
                activity.setSupportActionBar(toolbar.getToolbar());
                if (activity.getSupportActionBar() != null) {
                    activity.getSupportActionBar().setDisplayShowTitleEnabled(false);
                }
            }
        }

        // 根据屏幕宽度计算列数，卡片宽度约 120dp
        android.util.DisplayMetrics metrics = getResources().getDisplayMetrics();
        int screenWidth = metrics.widthPixels;
        int maxWidth = (int) (600 * metrics.density);
        int desiredCardWidth;
        if (screenWidth > maxWidth) {
            // 平板设备：卡片宽度 240dp
            desiredCardWidth = (int) (240 * metrics.density);
        } else {
            // 手机设备：卡片宽度约 120dp
            desiredCardWidth = (int) (120 * metrics.density);
        }
        int spanCount = Math.max(2, screenWidth / desiredCardWidth);
        
        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), spanCount));
        adapter = new FavoriteAdapter();
        adapter.setOnMovieClickListener(this);
        adapter.setOnMovieLongClickListener(this);
        recyclerView.setAdapter(adapter);

        btnDelete.setOnClickListener(v -> deleteSelected());

        etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                filterFavorites(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }

    private void initData() {
        repository = new MovieRepository(requireContext());
        loadFavorites();
    }

    private void loadFavorites() {
        new Thread(() -> {
            allFavorites = repository.getFavoriteMovies();
            requireActivity().runOnUiThread(() -> {
                adapter.setMovies(allFavorites);
                updateEmptyView();
            });
        }).start();
    }

    private void filterFavorites(String query) {
        if (query == null || query.trim().isEmpty()) {
            adapter.setMovies(allFavorites);
        } else {
            List<Movie> filtered = new ArrayList<>();
            String lowerQuery = query.toLowerCase(Locale.ROOT);
            for (Movie movie : allFavorites) {
                if (movie.getTitle().toLowerCase(Locale.ROOT).contains(lowerQuery)) {
                    filtered.add(movie);
                }
            }
            adapter.setMovies(filtered);
        }
        updateEmptyView();
    }

    private void updateEmptyView() {
        if (adapter.getItemCount() == 0) {
            emptyView.setVisibility(View.VISIBLE);
            recyclerView.setVisibility(View.GONE);
        } else {
            emptyView.setVisibility(View.GONE);
            recyclerView.setVisibility(View.VISIBLE);
        }
    }

    private void deleteSelected() {
        Set<String> selectedIds = adapter.getSelectedIds();
        if (selectedIds.isEmpty()) {
            Toast.makeText(getContext(), "请选择要删除的收藏", Toast.LENGTH_SHORT).show();
            return;
        }

        new androidx.appcompat.app.AlertDialog.Builder(getContext())
                .setTitle("删除收藏")
                .setMessage("确定要删除选中的 " + selectedIds.size() + " 部电影吗？")
                .setPositiveButton("删除", (dialog, which) -> {
                    new Thread(() -> {
                        for (String movieId : selectedIds) {
                            repository.removeFavorite(movieId);
                        }
                        requireActivity().runOnUiThread(() -> {
                            allFavorites.removeIf(m -> selectedIds.contains(m.getId()));
                            adapter.setSelectionMode(false);
                            adapter.setMovies(allFavorites);
                            bottomActions.setVisibility(View.GONE);
                            updateEmptyView();
                            Toast.makeText(getContext(), "已删除 " + selectedIds.size() + " 部收藏", Toast.LENGTH_SHORT).show();
                        });
                    }).start();
                })
                .setNegativeButton("取消", null)
                .show();
    }

    @Override
    public void onMovieClick(Movie movie, int position) {
        if (getActivity() instanceof MainActivity) {
            ((MainActivity) getActivity()).openDetail(movie.getId());
        }
    }

    @Override
    public boolean onMovieLongClick(Movie movie, int position) {
        if (!adapter.isSelectionMode()) {
            adapter.setSelectionMode(true);
            adapter.toggleSelection(movie.getId());
            bottomActions.setVisibility(View.VISIBLE);
        }
        return true;
    }

    @Override
    public void onResume() {
        super.onResume();
        loadFavorites();
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (!hidden) {
            loadFavorites();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (repository != null) {
            repository.close();
        }
    }
}

File: app/src/main/java/com/example/nasmovie/ui/FavoritesFragment.kt
package com.example.nasmovie.ui

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import com.example.nasmovie.R
import com.example.nasmovie.data.model.Movie
import com.example.nasmovie.databinding.FragmentFavoritesBinding
import com.example.nasmovie.ui.adapter.FavoriteAdapter

/**
 * 收藏 Fragment
 */
class FavoritesFragment : Fragment(),
    FavoriteAdapter.OnMovieClickListener,
    FavoriteAdapter.OnMovieLongClickListener {

    private var _binding: FragmentFavoritesBinding? = null
    private val binding get() = _binding!!

    private lateinit var adapter: FavoriteAdapter
    private lateinit var viewModel: FavoritesViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentFavoritesBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        viewModel = ViewModelProvider(this)[FavoritesViewModel::class.java]
        initViews(view)
        observeViewModel()
    }

    private fun initViews(view: View) {
        // 设置 Toolbar
        binding.toolbar?.let { toolbar ->
            toolbar.setTitle("我的收藏")
            toolbar.setShowBack(false)
            (activity as? AppCompatActivity)?.let { activity ->
                activity.setSupportActionBar(toolbar.toolbar)
                activity.supportActionBar?.setDisplayShowTitleEnabled(false)
            }
        }

        // 根据屏幕宽度计算列数，卡片宽度约 120dp
        val metrics = resources.displayMetrics
        val screenWidth = metrics.widthPixels
        val maxWidth = (600 * metrics.density).toInt()
        val desiredCardWidth = if (screenWidth > maxWidth) {
            // 平板设备：卡片宽度 240dp
            (240 * metrics.density).toInt()
        } else {
            // 手机设备：卡片宽度约 120dp
            (120 * metrics.density).toInt()
        }
        val spanCount = maxOf(2, screenWidth / desiredCardWidth)

        binding.recyclerViewFavorites.layoutManager = GridLayoutManager(context, spanCount)
        adapter = FavoriteAdapter()
        adapter.setOnMovieClickListener(this)
        adapter.setOnMovieLongClickListener(this)
        binding.recyclerViewFavorites.adapter = adapter

        binding.btnDelete.setOnClickListener { deleteSelected() }

        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                viewModel.filterFavorites(s.toString())
            }
            override fun afterTextChanged(s: Editable?) {}
        })
    }

    private fun observeViewModel() {
        viewModel.displayedFavorites.observe(viewLifecycleOwner) { favorites ->
            adapter.setMovies(favorites)
            updateEmptyView()
        }
    }

    private fun updateEmptyView() {
        if (adapter.itemCount == 0) {
            binding.emptyView.visibility = View.VISIBLE
            binding.recyclerViewFavorites.visibility = View.GONE
        } else {
            binding.emptyView.visibility = View.GONE
            binding.recyclerViewFavorites.visibility = View.VISIBLE
        }
    }

    private fun deleteSelected() {
        val selectedIds = adapter.selectedIds
        if (selectedIds.isEmpty()) {
            Toast.makeText(context, "请选择要删除的收藏", Toast.LENGTH_SHORT).show()
            return
        }

        AlertDialog.Builder(requireContext())
            .setTitle("删除收藏")
            .setMessage(getString(R.string.delete_favorites_confirm, selectedIds.size))
            .setPositiveButton("删除") { _, _ ->
                viewModel.deleteSelected(selectedIds)
                adapter.setSelectionMode(false)
                binding.bottomActions.visibility = View.GONE
                Toast.makeText(context, "已删除 ${selectedIds.size} 部收藏", Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton("取消", null)
            .show()
    }

    override fun onMovieClick(movie: Movie, position: Int) {
        (activity as? MainActivity)?.openDetail(movie.id)
    }

    override fun onMovieLongClick(movie: Movie, position: Int): Boolean {
        if (!adapter.isSelectionMode) {
            adapter.setSelectionMode(true)
            adapter.toggleSelection(movie.id)
            binding.bottomActions.visibility = View.VISIBLE
        }
        return true
    }

    override fun onResume() {
        super.onResume()
        viewModel.loadFavorites()
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) {
            viewModel.loadFavorites()
        }
    }
}